src/blog/articles/javascript-async-iterators/index.md

```diff
@@ -0,0 +1,689 @@
+    next () {
+      if (nextVal < 0) {
+        return { done: true }
+      }
+
+      return { 
+        done: false,
+        value: nextVal--
+      }
+    }
+  }
+}
+```
+
+In this example, `createCountdown` is a simple factory function. From this function, we return an _iterator_ object. In fact, the object implements the _iterator protocol_ defined above. Note that the returned object implements a `next()` method and that this method returns either `{done: true}` or `{done: false, value: someNumber}`.
+
+Let's see now how can we use this object to extract all the values:
+
+```javascript
+const countdown = createCountdown(3)
+console.log(countdown.next()) // { done: false, value: 3 }
+console.log(countdown.next()) // { done: false, value: 2 }
+console.log(countdown.next()) // { done: false, value: 1 }
+console.log(countdown.next()) // { done: false, value: 0 }
+console.log(countdown.next()) // { done: true }
+```
+
+Or if we want to use this object with a more generic loop:
+
+```javascript
+const countdown = createCountdown(3)
+let result = countdown.next()
+while (!result.done) {
+  console.log(result.value)
+  result = countdown.next()
+}
+```
+
+This will produce the following output:
+
+```text
+3
+2
+1
+0
+```
+
+This is not the most intuitive or ergonomic approach, but the iterator protocol is the basic building block for the _iterable protocol_ which enables the `for...of` syntax.
+
+## The JavaScript iterable protocol
+
+As we said, the _iterable protocol_ builds on top of the _iterator protocol_ that we just explored. Let's see how:
+
+> An object is **iterable** if it implements the `@@iterator` method, a zero-argument function that **returns an iterator**.
+
+Note that with `@@iterator` we indicate a symbol that is accessed with the global value `Symbol.iterator`.
+
+Can we make our countdown example _iterable_? We certainly can!
+
+```javascript
+function createCountdown (from) {
+  let nextVal = from
+  return {
+    [Symbol.iterator]: () => ({
+      next () {
+        if (nextVal < 0) {
+          return { done: true }
+        }
+
+        return { done: false, value: nextVal-- }
+      }
+    })
+  }
+}
+```
+
+In this new example, our factory function returns an _iterable_ object. This object in fact has a method referenced with `Symbol.iterator` that returns an _iterator_ object.
+
+At this point, once we have an instance of a countdown, we can use the `for..of` syntax to iterate over all the numbers from the countdown:
+
+```javascript
+const countdown = createCountdown(3)
+
+for (const value of countdown) {
+  console.log(value)
+}
+```
+
+This will output:
+
+```text
+3
+2
+1
+0
+```
+
+Hooray! Now we know how to make iterators and iterable objects. If you find the two terms confusing, don't worry, that's quite common!
+
+One good way to try to remember and distinguish these 2 concepts is the following:
+
+  - An _iterator_ is a lower-level object that allows us to iterate over some data set using `next()`
```

Comment:
An alternative definition (just a proposal and to provide more ideas)
- An iterator defines a protocol to iterate over an object
- An iterable defines a protocol to create iterators. Iterable objects can be used with for..of